core/src/test/java/kafka/test/api/ShareConsumerTest.java

```diff
@@ -0,0 +1,1651 @@
+        Properties props = cluster.clientProperties();
+        return new KafkaProducer<>(props, keySerializer, valueSerializer);
+    }
+
+    private <K, V> KafkaProducer<K, V> createProducer(Serializer<K> keySerializer,
+                                                      Serializer<V> valueSerializer,
+                                                      String transactionalId) {
+        Properties props = cluster.clientProperties();
+        props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId);
+        return new KafkaProducer<>(props, keySerializer, valueSerializer);
+    }
+
+    private <K, V> KafkaShareConsumer<K, V> createShareConsumer(Deserializer<K> keyDeserializer,
+                                                                Deserializer<V> valueDeserializer,
+                                                                String groupId) {
+        Properties props = cluster.clientProperties();
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+        return new KafkaShareConsumer<>(props, keyDeserializer, valueDeserializer);
+    }
+
+    private <K, V> KafkaShareConsumer<K, V> createShareConsumer(Deserializer<K> keyDeserializer,
+                                                                Deserializer<V> valueDeserializer,
+                                                                String groupId,
+                                                                Map<?, ?> additionalProperties) {
+        Properties props = cluster.clientProperties();
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+        props.putAll(additionalProperties);
+        return new KafkaShareConsumer<>(props, keyDeserializer, valueDeserializer);
+    }
+
+    private void warmup() throws InterruptedException, ExecutionException, TimeoutException {
+        createTopic(warmupTp.topic());
+        TestUtils.waitForCondition(() ->
+                        !scala.collection.JavaConverters.seqAsJavaList(cluster.brokers().get(0).metadataCache().getAliveBrokerNodes(new ListenerName("EXTERNAL"))).isEmpty(),
+                DEFAULT_MAX_WAIT_MS, 100L, () -> "cache not up yet");
+        ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(warmupTp.topic(), warmupTp.partition(), null, "key".getBytes(), "value".getBytes());
+        KafkaProducer<byte[], byte[]> producer = createProducer(new ByteArraySerializer(), new ByteArraySerializer());
+        producer.send(record).get(1000, TimeUnit.MILLISECONDS);
+        KafkaShareConsumer<byte[], byte[]> shareConsumer = createShareConsumer(new ByteArrayDeserializer(), new ByteArrayDeserializer(), "warmupgroup1");
+        Set<String> subscription = Collections.singleton(warmupTp.topic());
+        shareConsumer.subscribe(subscription);
+        TestUtils.waitForCondition(
+                () -> shareConsumer.poll(Duration.ofMillis(5000)).count() == 1, DEFAULT_MAX_WAIT_MS, 200L, () -> "warmup record not received");
```

Comment:
A few builds ago I see a timeout here https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-17009/4/testReport/junit/kafka.test.api/ShareConsumerTest/Build___JDK_8_and_Scala_2_12___testMultipleConsumersWithDifferentGroupIds__/

Is 15 seconds long enough to reliably see this record? We might need to increase this slightly.

The condition check is also pretty specific. Is more than one record possible? If so, we should relax the condition. If not, we should throw a more specific error.

Reply:
1. Yeah so first when it was failing, we were not waiting for the producer.send() future to complete. Now we are waiting for 10 seconds there, which solved the problems and the tests pass reliably.
2. Yes, only record is possible as we are producing 1 and trying to consume the same from the share consumer.